src/JSON/Deserializer.dfy

```diff
@@ -0,0 +1,154 @@
+include "Utils/Vectors.dfy"
+include "Utils/Unicode.dfy"
+include "Errors.dfy"
+include "AST.dfy"
+include "Grammar.dfy"
+include "Spec.dfy"
+
+module {:options "-functionSyntax:4"} JSON.Deserializer {
+  import Seq
+
+  import opened Wrappers
+  import opened BoundedInts
+  import opened Logarithm
+  import opened Power
+  import opened Utils.Str
+  import Utils.Unicode
+
+  import AST
+  import Spec
+  import opened Errors
+  import opened Utils.Vectors
+  import opened Grammar
+  import opened Utils.Views.Core
+
+  function Bool(js: Grammar.jbool): bool {
+    assert js.Bytes() in {Grammar.TRUE, Grammar.FALSE};
+    js.At(0) == 't' as byte
+  }
+
+  // TODO: Verify this function
+  function Unescape(str: string, start: nat := 0): DeserializationResult<string>
+    decreases |str| - start
+  { // Assumes UTF-16 strings
+    if start >= |str| then Success([])
+    else if str[start] == '\\' then
+      if |str| == start + 1 then
+        Failure(EscapeAtEOS)
+      else
+        var c := str[start + 1];
+        if c == 'u' then
+          if |str| <= start + 6 then
+            Failure(EscapeAtEOS)
+          else
+            var code := str[start + 2..start + 6];
+            if exists c | c in code :: c !in Str.HEX_TABLE then
+              Failure(UnsupportedEscape(code))
+            else
+              var tl :- Unescape(str, start + 6);
+              var hd := Str.ToNat(code, 16);
+              assert hd < 0x10000 by { reveal Pow(); }
+              Success([hd as char] + tl)
+        else
+          var unescaped: uint16 := match c
+             case '\"' => 0x22 as uint16 // quotation mark
+             case '\\' => 0x5C as uint16 // reverse solidus
+             case 'b'  => 0x08 as uint16 // backspace
+             case 'f'  => 0x0C as uint16 // form feed
+             case 'n'  => 0x0A as uint16 // line feed
+             case 'r'  => 0x0D as uint16 // carriage return
+             case 't'  => 0x09 as uint16 // tab
+             case _    => 0 as uint16;
+          if unescaped == 0 as uint16 then
+            Failure(UnsupportedEscape(str[start..start+2]))
+          else
+            var tl :- Unescape(str, start + 2);
+            Success([unescaped as char] + tl)
+    else
+      var tl :- Unescape(str, start + 1);
+      Success([str[start]] + tl)
+  }
+
+  function Transcode8To16Unescaped(str: seq<byte>): DeserializationResult<string>
+    // TODO Optimize with a function by method
+  {
+    Unescape(Unicode.Transcode8To16(str))
+  }
+
+  function String(js: Grammar.jstring): DeserializationResult<string> {
+    Transcode8To16Unescaped(js.contents.Bytes())
+  }
+
+  module ByteStrConversion refines Str.ParametricConversion {
+    import opened BoundedInts
+    type Char = byte
+  }
+
+  const DIGITS := map[
+    '0' as uint8 := 0, '1' as uint8 := 1, '2' as uint8 := 2, '3' as uint8 := 3,
+    '4' as uint8 := 4, '5' as uint8 := 5, '6' as uint8 := 6, '7' as uint8 := 7,
+    '8' as uint8 := 8, '9' as uint8 := 9
+  ]
```

Comment:
FYI, you could have written

map i <- "0123456789" :: i as uint8 := i

Reply:
Almost!  There would have to be a bit more arithmetic (`i as uint8` gives you the ascii code, so 49 for `1`, 50 for `2`, etc.)